PR: Show internal errors in a QMessageBox

rlaverde opened the pull request:
Fixes: #1525 

~~This isn't ready, I implemented but the behavior isn't the expected, the QMessageBox is shown but for every single line of the traceback (because `traceback_available` signal is emitted for every line).~~

~~any ideas of how I could fix this?~~

I implemented using a QMessageBox with non modal behavior

![traceback_box](https://cloud.githubusercontent.com/assets/2024217/22275240/32ca8e12-e279-11e6-887d-cf1a81792624.png)

goanpeca commented:
I dont think a QMessageBox will do then, you could create a custom dialog with a QTableView or something that would keep growing as more messages keep coming.

rlaverde commented:
@goanpeca Following your idea I found a way QMessageBox could be updated continuously :smile:

goanpeca commented:
:+1:

goanpeca commented:
@ccordoba12 do we want rich text on the details or not? and should they be links to the actual files?

Do we want to add a submit button so users can submit that error directly?

With that is that the issue tracker can start to grow a lot, which is good and bad at the same time!

ccordoba12 commented:
> do we want rich text on the details or not? and should they be links to the actual files?

Perhaps not because that could confuse users. Besides, clickable links are already available in the internal console and that's mostly useful for us ;-)

What we could have is highlighted tracebacks, but we could leave that for another PR.

> Do we want to add a submit button so users can submit that error directly?

Yep, that's really important to have ;-)

rlaverde commented:
I added the submit button, and rebase it

![spectacle m19702](https://cloud.githubusercontent.com/assets/2024217/26080140/d3f8f0d6-398a-11e7-9eb0-2ae97be09c48.png)

ccordoba12 commented on spyder/widgets/sourcecode/codeeditor.py:
```diff
@@ -2425,6 +2425,7 @@ def __get_current_color(self, cursor=None):
 
     def in_comment_or_string(self, cursor=None):
         """Is the cursor inside or next to a comment or string?"""
+        raise Exception
```
This is making our tests to fail. It seems an artifact of the rebase, so please remove it.

rlaverde replied:
Ohhh! sorry, that was for testing, I didn't meant to commit that 🙈

ccordoba12 commented:
@rlaverde, could use an instance of CodeEditor to show the tracebacks? I think that would make the dialog more useful :-)

rlaverde commented:
> Could we use an instance of CodeEditor to show the tracebacks? I think that would make the dialog more useful :-)

I used a customized QMessageBox, for embedding the CodeEditor I will have to rewrite it using a Dialog and implement the layout and functionality

goanpeca commented:
>@rlaverde, could we use an instance of CodeEditor to show the tracebacks? I think that would make the dialog more useful :-)

@ccordoba12 I would say for now this is good enough and as @rlaverde he would have to basically redo all his PR for what you are asking

ccordoba12 commented:
> I would say for now this is good enough and as @rlaverde he would have to basically redo all his PR for what you are asking

Ok, no problem, let's leave things as they are.

One more thing: we need to preserve the old behavior when DEV is active. That's way more practical for us developers because traceback links are clickable in the Internal console.

rlaverde commented:
> One more thing: we need to preserve the old behavior when DEV is active. That's way more practical for us developers because traceback links are clickable in the Internal console.

Done, also when `DEBUG` or `PYTEST` is activated

ccordoba12 commented on spyder/plugins/console.py:
```diff
@@ -197,13 +202,44 @@ def register_plugin(self):
+                self.msgbox_traceback = QMessageBox(
+                    QMessageBox.Critical,
+                    _('Error'),
+                    _("<b>Spyder-IDE has encountered a problem.</b>"
```
Remove the `-IDE` here.

ccordoba12 replied:
Also add a `<br>` at the end.

ccordoba12 commented on spyder/plugins/console.py:
```diff
@@ -197,13 +202,44 @@ def register_plugin(self):
+                      "Sorry for the inconvenience."
+                      "<b>Please tell us about this problem.</b>"
+                      "<br><br>"
+                      "You can submit this error to the github issue tracker"),
```
Let's change this to

> You can automatically submit this error to our Github issues tracker.<br>
*Note*: You need a Github account for that.

ccordoba12 commented on spyder/plugins/console.py:
```diff
@@ -197,13 +202,44 @@ def register_plugin(self):
+                    parent=self)
+
+                self.submit_btn = self.msgbox_traceback.addButton(
+                        _('Submit to github'), QMessageBox.YesRole)
```
`github` -> `Github`

ccordoba12 commented on spyder/plugins/console.py:
```diff
@@ -197,13 +202,44 @@ def register_plugin(self):
+                    _('Error'),
+                    _("<b>Spyder-IDE has encountered a problem.</b>"
+                      "Sorry for the inconvenience."
+                      "<b>Please tell us about this problem.</b>"
```
Let's remove this line.

ccordoba12 commented:
I added some minor formatting comments. After that, I think this is ready!

rlaverde commented:
I just addressed your comments, this is how it looks like:

![spectacle jg8171](https://user-images.githubusercontent.com/2024217/27105732-410550fc-5056-11e7-96dd-036b405b870d.png)

ccordoba12 commented:
I like it a lot, thanks!

@goanpeca, what do you say?
